pkg/gadgets/socket-collector/gadget.go

```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
-		}
-		if len(trace.Spec.Filter.Labels) != 0 {
-			gadgets.SetStatusError(trace, "TODO: Filtering by labels is not currently supported")
-			return
-		}
 		if trace.Spec.Filter.ContainerName != "" {
-			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!", trace.Spec.Gadget)
+			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!",
+				trace.Spec.Gadget)
 		}
 
-		pidsInPod := t.resolver.LookupPIDByPod(trace.Spec.Filter.Namespace, trace.Spec.Filter.Podname)
-
-		// All containers inside a pod share the same network namespace
-		// thus we can just take the first valid pid we found
-		for _, pid = range pidsInPod {
-			if pid != 0 {
-				break
-			}
+		selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
+		filteredContainers := t.resolver.GetContainersBySelector(selector)
+		if len(filteredContainers) == 0 {
+			gadgets.SetStatusError(trace, fmt.Sprintf("No container matches the requested filter: %+v",
+				*trace.Spec.Filter))
+			return
 		}
 
-		log.Infof("Gadget %s: Using PID %d to retrieve network namespaces of Podname %q in Namespace %q",
-			trace.Spec.Gadget, pid, trace.Spec.Filter.Podname, trace.Spec.Filter.Namespace)
-
-		if pid == 0 {
-			gadgets.SetStatusError(trace, fmt.Sprintf("Couldn't find a valid PID for Podname %q in Namespace %q",
-				trace.Spec.Filter.Podname, trace.Spec.Filter.Namespace))
+		output, err := start(trace.Spec.Gadget, trace.Spec.Node, filteredContainers)
+		if err != nil {
+			gadgets.SetStatusError(trace, err.Error())
 			return
 		}
 
-		t.Start(trace, pid)
+		trace.Status.OperationError = ""
+		trace.Status.Output = output
+		trace.Status.State = "Completed"
 	default:
 		gadgets.SetStatusError(trace, fmt.Sprintf("Unknown operation %q", operation))
 	}
 }
 
-func (t *Trace) Start(trace *gadgetv1alpha1.Trace, pid uint32) {
-	output, err := tracer.RunCollector(
-		pid,
-		trace.Spec.Filter.Podname,
-		trace.Spec.Filter.Namespace,
-		trace.Spec.Node,
-	)
+func start(gadgetName, node string, filteredContainers []pb.ContainerDefinition) (string, error) {
+	allSockets := []socketcollectortypes.Event{}
+
+	// Given that the socket-collector tracer works per network namespace and
+	// all the containers inside a namespace/pod share the network namespace,
+	// we only need to run the tracer with one valid PID per namespace/pod
+	visitedPods := make(map[string]bool)
+
+	for _, container := range filteredContainers {
+		key := container.Namespace + "/" + container.Podname
+		if _, ok := visitedPods[key]; !ok {
+			// Make the whole gadget fail if there is a container without PID
+			// because it would be an inconsistency that has to be notified
+			if container.Pid == 0 {
+				return "", fmt.Errorf("aborting! The following container does not have PID %+v", container)
+			}
+
+			// The stored value does not matter, we are just keeping
+			// track of the visited Pods per Namespace
+			visitedPods[key] = true
```

Comment:
This is more Go idiomatic:

```suggestion
			visitedPods[key] = struct{}{}
```

Reply:
TIL:
> The struct{} is a struct type with zero elements. It is often used when no information is to be stored. It has the benefit of being 0-sized, so usually no memory is required to store a value of type struct{}. [Ref](https://stackoverflow.com/a/45123114).

Applied. Thanks!